### PR TITLE
feat(medusa,core-flows,types): add cart <> tax integration workflows + steps

### PR DESCRIPTION
**Pull request by riqwan**
what:

- adds tax lines to cart when item operations take place

RESOLVES CORE-1821
RESOLVES CORE-1822
RESOLVES CORE-1823
RESOLVES CORE-1824

### CODE REVIEW

File: packages/core-flows/src/definition/cart/steps/set-tax-lines-for-items.ts [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    await cartService.setLineItemTaxLines(
+      cart.id,
+      normalizeItemTaxLinesForCart(lineItemTaxLines)
+    )
+
+    await cartService.setShippingMethodTaxLines(
+      cart.id,
+      normalizeShippingTaxLinesForCart(shippingMethodTaxLines)
+    )
```
Comment (srindom) [inline packages/core-flows/src/definition/cart/steps/set-tax-lines-for-items.ts]: suggestion: parallelize these

File: packages/core-flows/src/definition/cart/workflows/create-carts.ts [hunk 2]
```diff
@@ -127,9 +132,16 @@ export const createCartWorkflow = createWorkflow(
...
     const carts = createCartsStep([cartToCreate])
-
     const cart = transform({ carts }, (data) => data.carts?.[0])
 
+    updateTaxLinesStep(
+      transform({ cart }, (data) => {
+        return {
+          cartOrCartId: data.cart.id,
+        }
+      })
```
Comment (srindom) [inline packages/core-flows/src/definition/cart/workflows/create-carts.ts]: q: is the transform necessary?

Comment (riqwan) [reply packages/core-flows/src/definition/cart/workflows/create-carts.ts]: Nope, not necessary. Removed it 🤞🏻

File: integration-tests/modules/__tests__/cart/store/carts.spec.ts [hunk 3]
```diff
@@ -682,17 +873,45 @@ describe("Store Carts API", () => {
           ]),
         })
       )
+
+      response = await api.post(`/store/carts/${cart.id}/line-items`, {
+        variant_id: productWithDefaultTax.variants[0].id,
+        quantity: 1,
+      })
+
+      expect(response.data.cart).toEqual(
```
Comment (srindom) [inline integration-tests/modules/__tests__/cart/store/carts.spec.ts]: suggestion: should we check the total here to see if we get it right?

File: packages/core-flows/src/definition/cart/steps/update-tax-lines.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+import { updateTaxLinesWorkflow } from "../workflows"
+
+interface StepInput {
+  cart_or_cart_id: CartWorkflowDTO | string
+  items?: CartLineItemDTO[]
+  shipping_methods?: CartShippingMethodDTO[]
+}
+
+export const updateTaxLinesStepId = "update-tax-lines-step"
+export const updateTaxLinesStep = createStep(
```
Comment (srindom) [inline packages/core-flows/src/definition/cart/steps/update-tax-lines.ts:16]: thought: this should potentially have a check to see if this is an automatic tax region.

**Review by srindom — APPROVED** [review]
LGTM - still need work for automatic vs. manual tax but can come in separate PR

File: integration-tests/modules/__tests__/cart/store/carts.spec.ts (hunk 3 above)
Comment (riqwan) [reply integration-tests/modules/__tests__/cart/store/carts.spec.ts]: totals aren't done in the cart APIs or module yet, afaik. Or are you talking about something else?

File: packages/core-flows/src/definition/cart/steps/update-tax-lines.ts (hunk 4 above)
Comment (riqwan) [reply packages/core-flows/src/definition/cart/steps/update-tax-lines.ts:16]: will bring this over to the next PR

File: packages/core-flows/src/definition/cart/workflows/create-carts.ts [hunk 5]
```diff
@@ -83,26 +84,30 @@ export const createCartWorkflow = createWorkflow(
...
-          "product.thumbnail",
-          "product.type",
-          "product.collection",
-          "product.handle",
-        ],
-        relations: ["product"],
-      },
-    })
+    const variants = getVariantsStep(
+      transform({ variantIds }, (data) => {
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/workflows/create-carts.ts:88]: question: What happened here?

Comment (riqwan) [reply packages/core-flows/src/definition/cart/workflows/create-carts.ts:88]: Interesting bug that I came across. If you don't use the transform here, you get data from an earlier workflow run. I have a task on my to-do to check this behavior.

Comment (adrien2p) [reply packages/core-flows/src/definition/cart/workflows/create-carts.ts:88]: I ve fixed it yesterday 😊 oli had the same issue

Comment (olivermrbl) [reply packages/core-flows/src/definition/cart/workflows/create-carts.ts:88]: Got it. I believe this was solved in 163bd0207cadb647bb94122d85d266ee29547495.

I'll revert and give it a spin.

Comment (riqwan) [reply packages/core-flows/src/definition/cart/workflows/create-carts.ts:88]: Nice! 🎉

**Pull request merged**